src/shogun/io/ARFFFile.cpp

```diff
@@ -0,0 +1,363 @@
+					if (type.size() == 1)
+						m_date_formats.emplace_back(m_default_date_format);
+					else
+						m_date_formats.push_back(
+						    javatime_to_cpptime(date_elements[1]));
+					name = "";
+				}
+				else if (date_elements[1] == "date" && date_elements.size() < 4)
+				{
+					// @attribute name date [[date-format]]
+					if (date_elements.size() == 2)
+						m_date_formats.emplace_back(m_default_date_format);
+					else
+						m_date_formats.push_back(
+						    javatime_to_cpptime(date_elements[2]));
+				}
+				else
+				{
+					SG_SERROR(
+					    "Error parsing date on line %d: %s\n", m_line_number,
+					    m_current_line.c_str())
+				}
+				m_attributes.push_back(Attribute::Date);
+				m_data_vectors.emplace_back(std::vector<float64_t>{});
+			}
+			else if (is_primitive_type(type))
+			{
+				type = to_lower(type);
+				// numeric attributes
+				if (type == "numeric")
+				{
+					m_attributes.push_back(Attribute::Numeric);
+					m_data_vectors.emplace_back(std::vector<float64_t>{});
+				}
+				else if (type == "integer")
+				{
+					m_attributes.push_back(Attribute::Integer);
+					m_data_vectors.emplace_back(std::vector<float64_t>{});
+				}
+				else if (type == "real")
+				{
+					m_attributes.push_back(Attribute::Real);
+					m_data_vectors.emplace_back(std::vector<float64_t>{});
+				}
+				else if (type == "string")
+				{
+					// @ATTRIBUTE LCC    string
+					m_attributes.push_back(Attribute::String);
+					m_data_vectors.emplace_back(std::vector<std::string>{});
+				}
+				else
+					SG_SERROR(
+					    "Unexpected attribute type identifier \"%s\" "
+					    "on line %d: %s\n",
+					    type.c_str(), m_line_number, m_current_line.c_str())
+			}
+			else
+				SG_SERROR(
+				    "Unexpected format in @ATTRIBUTE on line %d: %s\n",
+				    m_line_number, m_current_line.c_str());
+			m_attribute_names.emplace_back(name);
+		}
+		// comments in this section are ignored
+		else if (m_current_line.substr(0, 1) == m_comment_string)
+		{
+		}
+		else if (
+		    to_lower(m_current_line.substr(0, strlen(m_data_string))) ==
+		    m_data_string)
+			m_state = true;
+	};
+
+	auto check_attributes = [this]() {
+		// attributes cannot be empty
+		return !m_attributes.empty();
+	};
+	process_chunk(read_attributes, check_attributes, true);
+
+	auto pos = m_stream->tellg();
+	auto approx_data_line_count = std::count(std::istreambuf_iterator<char>(*m_stream),
+	        std::istreambuf_iterator<char>(), '\n');
+	reserve_vector_memory(approx_data_line_count);
+	m_stream->seekg(pos);
+
+	auto read_data = [this]() {
+		// it's a comment and can be skipped
+		if (SG_UNLIKELY(m_current_line.substr(0, 1) == m_comment_string))
+			return;
+		// it's the data string (i.e. @data"), does not provide information
+		if (SG_UNLIKELY(
+		        to_lower(m_current_line.substr(0, strlen(m_data_string))) ==
+		        m_data_string))
+		{
+			return;
+		}
+		// assumes that until EOF we should expect comma delimited values
```

Comment:
maybe a check with a user facing error msg since users might screw that up?

Reply:
I put a check for element count that should catch any weird behaviour here